client/my-sites/site-settings/site-icon-setting/index.jsx

```diff
@@ -40,11 +51,45 @@ class SiteIconSetting extends Component {
 
 	showModal = () => this.toggleModal( true );
 
-	setSiteIcon = ( media ) => {
-		// [TODO]: Handle setting site icon
-		console.log( media ); // eslint-disable-line no-console
+	editSelectedMedia = ( media ) => {
+		if ( media ) {
+			this.props.onEditSelectedMedia();
+		}
+	};
 
+	setSiteIcon = ( error, blob ) => {
 		this.hideModal();
+		this.props.resetAllImageEditorState();
+
+		if ( error || ! blob ) {
+			return;
+		}
+
+		const { siteId } = this.props;
+		const selectedItem = head( MediaLibrarySelectedStore.getAll( siteId ) );
+		if ( ! selectedItem ) {
+			return;
+		}
+
+		const transientMediaId = uniqueId( 'site-icon' );
+
+		MediaActions.add( siteId, {
+			ID: transientMediaId,
+			fileName: `cropped-${ selectedItem.file }`,
+			fileContents: blob
+		} );
+
+		const checkUploadComplete = () => {
+			const transientMedia = MediaStore.get( siteId, transientMediaId );
+			if ( isItemBeingUploaded( transientMedia ) ) {
+				return;
+			}
+
+			MediaStore.off( 'change', checkUploadComplete );
+			alert( transientMedia.ID );
+		};
+
+		MediaStore.on( 'change', checkUploadComplete );
```

Comment:
I'm wondering if we could have a "transition" state, where we would move the lines from 74:92 here to a redux thunk action creator. `MediaActions` will be used like an API abstraction.